cluster_experiments/experiment_analysis.py

```diff
@@ -301,6 +302,114 @@ def from_config(cls, config):
+
+    Arguments:
+        cluster_cols: list of columns to use as clusters
+        target_col: name of the column containing the variable to measure
+        treatment_col: name of the column containing the treatment variable
+        treatment: name of the treatment to use as the treated group
+        strata_cols: list of index columns for paired t test. Should be a subset or equal to cluster_cols
+
+    Usage:
+
+    ```python
+    from cluster_experiments.experiment_analysis import PairedTTestClusteredAnalysis
+    import pandas as pd
+
+    df = pd.DataFrame({
+        'x': [1, 2, 3, 4, 0, 0, 1, 1],
+        'treatment': ["A", "B", "A", "B"] * 2,
+        'cluster': [1, 2, 3, 4, 1, 2, 3, 4],
+    })
+
+    PairedTTestClusteredAnalysis(
+        cluster_cols='cluster',
+        strata_cols='cluster',
+        target_col=df['x'],
+    ).get_pvalue(df)
+    ```
+    """
+
+    def __init__(
+        self,
+        cluster_cols: List[str],
+        strata_cols: List[str],
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+    ):
+        self.strata_cols = strata_cols
+        self.target_col = target_col
+        self.treatment = treatment
+        self.treatment_col = treatment_col
+        self.cluster_cols = cluster_cols
+
+    def _preprocessing(self, df: pd.DataFrame) -> pd.DataFrame:
+        df_grouped = df.groupby(
+            self.cluster_cols + [self.treatment_col], as_index=False
+        )[self.target_col].mean()
+
+        n_control = df_grouped[self.treatment_col].value_counts()[0]
+        n_treatment = df_grouped[self.treatment_col].value_counts()[1]
+
+        if n_control != n_treatment:
+            logging.warning(
+                f"groups don't have same number of observations, {n_treatment =} and  {n_control =}"
+            )
+
+        assert all(
+            [x in self.cluster_cols for x in self.strata_cols]
+        ), f"strata should be a subset or equal to cluster_cols ({self.cluster_cols = }, {self.strata_cols = })"
+
+        df_pivot = df_grouped.pivot_table(
+            columns=self.treatment_col,
+            index=self.strata_cols,
+            values=self.target_col,
+        )
+
+        if df_pivot.isna().sum().sum() == 0:
+            logging.warning(
+                f"There are missing data from some clusters, removing them: {df_pivot[df_pivot.isna().any(axis=1)].to_dict()}"
+            )
+
+        df_pivot = df_pivot.dropna()
+
+        return df_pivot
+
+    def analysis_pvalue(self, df: pd.DataFrame, verbose: bool = False) -> float:
+        """Returns the p-value of the analysis
+        Arguments:
+            df: dataframe containing the data to analyze
+            verbose (Optional): bool, prints the regression summary if True
+        """
+
+        df_pivot = self._preprocessing(df=df)
+
+        if verbose:
+            print(f"performing paired t test in this data \n {df_pivot} \n")
```

Comment:
can you put this print below? so that we only have 1 if verbose, and decouple printing and calculation (I know it's a bit weird because you're lying). if you don't like it, I suggest to move this part of the code to _preprocessing